src/cir.rs

```diff
@@ -0,0 +1,437 @@
+use crate::ast::Predicate;
+use crate::context::{Context, Match};
+use crate::interpreter::Execute;
+use crate::lir::{is_operator, LirInstruction, LirLogicalOperators, LirProgram, Translate};
+use crate::semantics::FieldCounter;
+use std::collections::HashMap;
+
+#[derive(Debug)]
+pub struct CirProgram {
+    pub(crate) instructions: Vec<CirInstruction>,
+}
+
+impl CirProgram {
+    pub fn new() -> Self {
+        Self {
+            instructions: Vec::new(),
+        }
+    }
+}
+
+impl Default for CirProgram {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+#[derive(Debug)]
+pub enum CirInstruction {
+    AndIns(AndIns),
+    OrIns(OrIns),
+    NotIns(NotIns),
+}
+
+#[derive(Debug)]
+pub struct AndIns {
+    left: CirOperand,
+    right: CirOperand,
+}
+
+#[derive(Debug)]
+pub struct OrIns {
+    left: CirOperand,
+    right: CirOperand,
+}
+
+#[derive(Debug)]
+pub struct NotIns {
+    right: CirOperand,
+}
+
+#[derive(Debug, Clone)]
+pub enum CirOperand {
+    Index(usize),
+    Predicate(Predicate),
+}
+
+impl CirOperand {
+    pub fn as_predicate(&self) -> Option<&Predicate> {
+        match &self {
+            CirOperand::Index(_index) => None, // never be here, otherwise something wrong
+            CirOperand::Predicate(p) => Some(p),
+        }
+    }
+
+    pub fn as_index(&self) -> Option<usize> {
+        match &self {
+            CirOperand::Index(index) => Some(*index),
+            CirOperand::Predicate(_p) => None, // never be here, otherwise something wrong
+        }
+    }
+}
```

Comment:
I would like to suggest using `panic!` instead of `Option` and then `unwrap` on the use case if `None` is impossible case for that. Like:
```Rust
pub fn as_index(&self) -> usize {
    match &self {
        CirOperand::Index(index) => *index,
        CirOperand::Predicate(_) => panic!("Call as_index on predicate Operand, CirProgram is wrong.")
    }
}
```

This will eliminate verbose `unwrap` usage and make a better panic message.

Reply:
fixed, good point. thanks.